internals/overlord/logstate/clienterr/clienterr.go

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+/*
+Package clienterr contains common error types which are recognised by the log
+gatherer. logstate.logClient implementations should return these error types
+to communicate with the gatherer.
+*/
+package clienterr
+
+import (
+	"time"
+)
+
+// Backoff should be returned if the server indicates we are sending too many
+// requests (e.g. an HTTP 429 response).
+type Backoff struct {
+	RetryAfter *time.Time
+}
+
+func (e *Backoff) Error() string {
+	errStr := "too many requests"
+	if e.RetryAfter != nil {
+		errStr += ", retry after " + e.RetryAfter.String()
+	}
+	return errStr
+}
+
+// Timeout should be returned if the client's Flush times out.
+type Timeout struct {
```

Comment:
Given the content of the error string do we think this error is properly named? Should it be a flush timeout error?

Is is it worth making a quick `typedef ConstError string` somewhere here or around here and making Timeout a const error? This way the user that is decorating and existing error gets to decided on the wording of the error better?

Just some random thoughts :)

Reply:
all of these are implicitly flush errors. All a client does is `Write` and `Flush`, and the writing itself is probably never gonna cause an error.

Reply:
That is fine. Can we update the comment here?

I would argue that if the comment on the error says this is what gets returned when a flush error happens then it's name should reflect it's very specific purpose.

If it's just a Timeout error then just call it that and update the comment on the Flush interface to say that an error that satisfies Timeout will be returned.

It's very nitty but comments are making contracts with the user IMO so lets just be careful where we are putting specifics.

Reply:
I've now removed `Timeout` so this discussion is moot